feat: Add gas estimation utilities (#6267)
ethereum-optimism/optimism: packages/fee-estimation/README.md

> # GasEstimation Documentation

Author: ```suggestion
# @eth-optiism/fee-estimation Documentation
```

## Why

> Gas estimation on OP-chains has both an l2 and l1 component. By default tools such as Viem, Wagmi, Ethers, and Web3.js do not support the l1 component. They will support this soon but in meantime this library can help estimate fees for transactions.

Reviewer: ```suggestion
Fee estimation on OP-chains has both an l2 and l1 component. By default tools such as Viem, Wagmi, Ethers, and Web3.js do not support the l1 component. They will support this soon but in meantime, this library can help estimate fees for transactions, or act as a reference.
```

As these tools add support for gas estimation natively this README will be updated with framework specific instructions.

## High level

- The l2 contract that can estimate l1Fees is called [GasPriceOracle](../contracts-bedrock/contracts/l2/GasPriceOracle.sol) contract. This library provides utils for interacting with it at a high level.
- The GasPriceOracle is [deployed to Optimism](https://optimistic.etherscan.io/address/0x420000000000000000000000000000000000000F) and other OP chains at a predeployed address of `0x420000000000000000000000000000000000000F`

> For more detailed information about gas fees on Optimism's Layer 2, you can visit their [official documentation](https://community.optimism.io/docs/developers/build/transaction-fees/#the-l2-execution-fee).

Author: Should mention that the main method here is estimateFees since 9/10 folks using this library only care about that one function

Reviewer: agreed! calling at the function first and foremost

## Installation:

```bash
pnpm i @eth-optimism/fee-estimation
```

```bash
npm i @eth-optimism/fee-estimation
```

```
yarn add @eth-optimism/fee-estimation
```

## Importing the Gas Fee Estimator

To use the Gas Fee Estimator in your file, import it:

```typescript
import {
  baseFee,
  decimals,
  estimateFees,
  gasPrice,
  getL1Fee,
  getL1GasUsed,
  l1BaseFee,
  overhead,
  scalar,

>   version,

Author: Use vscode to sort this alphabetically

Author: gonna regenerate the entire readme and will make sure this is done at that time

> } from '@eth-optimism/fee-estimator'

Author: Looks like a mix up between fee-estimator and fee-estimation going on make sure these are consistent

Reviewer: not particularly opinionated, maybe fee-estimation?

Author: yea fee-estimation is better imo

Author: ```suggestion
} from '@eth-optimism/fee-estimation'
```

```

## Usage

### Provider Options

Provider options are required for most function calls. They specify the chain and RPC URL for fetching data. They can be defined as follows:

```typescript
const chainConfig = {
  chainId: 10,
  rpcUrl: 'https://mainnet.optimism.io',

> }

Reviewer: will people typically have something like this chain config lying around in the same file they make those calls?

Reviewer: can we make it possible or call out how to pass in stuff from an existing provider?

Author: Yea definitely can get this config off of an existing provider or if using wagmi/viem could import from `viem/chains`. Could even make the methods take an ethers provider or viem provider as an optional argument

Author: Correction. Able to offer passing in a viem client in but ethers.js provider instance makes the rpcUrl completely private https://github.com/ethers-io/ethers.js/blob/5a56fc36d7bd78ed155308ec7bd0283ece2f6dbd/src.ts/providers/provider-jsonrpc.ts#L1124.

Ethers users should have the url laying around somewhere so the documentation will tell them to create an `l2Client` with it and use that

```

### Block Options

Block options specify the block from which to retrieve data. They can be defined as follows:

```typescript
const blockConfig = {
  blockNumber: BigInt(106889079),

> }

Reviewer: same Q as above, not a blocker but just wondering if configuring this will mean more work on the average case

Author: This is an optional config nobody would usually use. This is only for when you are querying a past blockNumber.  I use it specifically in tests because using it in all my tests makes it so the tests are very deterministic unlikely to break in future since they always run against the same state.  99% of users would just leave this blank which tells viem to use the default `latest` blocktag

Reviewer: calling out the optional-ness would be nice! i thought it wasn't on first read

```

These options can then be combined to form the parameters for function calls:

```typescript
const params = { ...chainConfig, ...blockConfig }
```

### Retrieving Base Fee

To retrieve the base fee:

```typescript
const baseFeeValue = await baseFee(params)
```

### Retrieving Gas Price

To retrieve the gas price:

```typescript
const gasPriceValue = await gasPrice(params)
```

### Estimating Fees

To estimate fees for a transaction:

```typescript
const data = '0xd1e16f0a603ac...' // transaction data
const account = '0xe371815c5f8a4f9acd1576879de288acd81269f1'
const to = '0xe35f24470730f5a488da9721548c1ab0b65b53d5'
const fees = await estimateFees({ ...params, to, data, account })
```

> This will return the estimated gas fee for the specified transaction.

Reviewer: would love to see the expected type/format of what's returned

Author: Yea these docs look... very chat gpt. Definitely can do better

Author: Way better docs now

> # @eth-optimism/fee-estimation 

Author: These docs will have to get regenerated

This library provides a higher level abstraction over the gasPriceOracle

## Installation

```bash
pnpm install @eth-optimism/fee-estimation
```

```bash
npm install @eth-optimism/fee-estimation
```

```bash
yarn add @eth-optimism/fee-estimation
```

## Usage

### Import the package

```ts
import { estimateFees } from '@eth-optimism/fee-estimation';
```

### Basic Usage

```ts
const fees = await estimateFees({
  client: {
    chainId: 10,
    rpcUrl: 'https://mainnet.optimism.io',
  },
  blockNumber: BigInt(106889079),
  account: '0xe371815c5f8a4f9acd1576879de288acd81269f1',
  to: '0xe35f24470730f5a488da9721548c1ab0b65b53d5',
  data: "0x5c19a95c00000000000000000000000046abfe1c972fca43766d6ad70e1c1df72f4bb4d1",
});
```

## API

### `estimateFees` function

```ts
estimateFees(params: EstimateFeeParams): Promise<bigint>
```

#### Parameters

`params`: An object with the following fields:

- `client`: An object with `rpcUrl` and `chainId` fields, or an instance of a Viem `PublicClient`.

> - `blockNumber`: A BigInt representing the block number at which you want to estimate the fees.

Author: ```suggestion
- `blockNumber(optional)`: A BigInt representing the block number at which you want to estimate the fees.
```

> - `blockTag`: A string representing the block tag to query from.

Author: ```suggestion
- `blockTag(optional)`: A string representing the block tag to query from.
```

- `account`: A string representing the account making the transaction.

- `to`: A string representing the recipient of the transaction.

> - `data`: A string representing the data being sent with the transaction. This should be a 0x-prefixed hex string.

Author: ```suggestion
- `data`: A string representing the data being sent with the transaction. This should be a 0x-prefixed hex string. Can be generated with `encodeFunctionData`
```

Returns the decimals used in the scalar.

```ts
decimals({ client, ...params }: GasPriceOracleOptions): Promise<bigint>;
```

#### Parameters
- `{ client, ...params }: GasPriceOracleOptions` - The options required to fetch the decimals.

#### Returns
- `Promise<bigint>` - Returns a promise that resolves to the decimals used in the scalar.

#### Example

```ts
const decimalsValue = await decimals(paramsWithClient);
```

---

### gasPrice()

Returns the gas price.

```ts
gasPrice({ client, ...params }: GasPriceOracleOptions): Promise<bigint>;
```

#### Parameters
- `{ client, ...params }: GasPriceOracleOptions` - The options required to fetch the gas price.

#### Returns
- `Promise<bigint>` - Returns a promise that resolves to the gas price.

#### Example

```ts
const gasPriceValue = await gasPrice(paramsWithClient);
```

---

### getL1Fee()

Computes the L1 portion of the fee based on the size of the rlp encoded input transaction, the current L1 base fee, and the various dynamic parameters.

```ts
getL1Fee(data: Bytes, { client, ...params }: GasPriceOracleOptions): Promise<bigint>;
```

#### Parameters
- `data: Bytes` - The transaction call data as a 0x-prefixed hex string.

> - `{ client, ...params }: GasPriceOracleOptions` - Optional lock options and provider options.

Author: I didn't review these well can replace these with more explicit params. This is copy pasted from the tests

```ts
estimateFees({ client, data, account, to, blockNumber }: EstimateFeesParams): Promise<bigint>;
```

#### Parameters
- `{ client, data, account, to, blockNumber }: EstimateFeesParams` - An object containing the client, transaction data, sender's address, recipient's address, and block number.

#### Returns
- `Promise<bigint>` - Returns a promise that resolves to the estimated fee for the given transaction.

#### Example

```ts
const estimateFeesParams = {
  data: '0xd1e16f0a603acf1f8150e020434b096e408bafa429a7134fbdad2ae82a9b2b882bfcf5fe174162cf4b3d5f2ab46ff6433792fc99885d55ce0972d982583cc1e11b64b1d8d50121c0497642000000000000000000000000000000000000060a2c8052ed420000000000000000000000000000000000004234002c8052edba12222222228d8ba445958a75a0704d566bf2c84200000000000000000000000000000000000006420000000000000000000000000000000000004239965c9dab5448482cf7e002f583c812ceb53046000100000000000000000003',
  account: '0xe371815c5f8a4f9acd1576879de288acd81269f1',
  to: '0xe35f24470730f5a488da9721548c1ab0b65b53d5',
};
const estimatedFees = await estimateFees({ ...paramsWithClient, ...estimateFeesParams });
```

> I hope this information is helpful!

Author: ```suggestion
```
It was thank you chat gpt
